autogen/src/main.rs

```diff
@@ -76,16 +167,459 @@ async fn update_json() -> Result<()> {
+    .iter()
+    .map(|(k, v)| (k.to_lowercase().clone(), v.clone()))
+    .collect();
+  for (name, _) in schema.schema.object.unwrap().properties {
+    if let Some(schema) = defs1.get(name.as_str()).cloned() {
+      let schema = schema.into_object();
+      write_schema(
+        writer,
+        name,
+        schema,
+        &defs,
+        "FIELD_DEFINITION",
+        written_directives,
+        extra_it,
+      )?;
+    }
+  }
+
+  Ok(())
+}
+
+fn write_array_validation(
+  writer: &mut IndentedWriter<impl Write>,
+  name: String,
+  arr_valid: ArrayValidation,
+  defs: &BTreeMap<String, Schema>,
+  extra_it: &mut BTreeMap<String, ObjectValidation>,
+) -> std::io::Result<()> {
+  let mut is_required = false;
+  write!(writer, "[")?;
+  if let Some(SingleOrVec::Single(schema)) = arr_valid.items {
+    is_required = true;
+    write_type(writer, name, schema.into_object(), defs, extra_it)?;
+  } else if let Some(SingleOrVec::Vec(schemas)) = arr_valid.items {
+    write_type(writer, name, schemas[0].clone().into_object(), defs, extra_it)?;
+  } else {
+    println!("{name}: {arr_valid:?}");
+
+    write!(writer, "JSON")?;
+  }
+  if is_required {
+    write!(writer, "]!")
+  } else {
+    write!(writer, "]")
+  }
+}
+
+fn write_object_validation(
+  writer: &mut IndentedWriter<impl Write>,
+  name: String,
+  obj_valid: ObjectValidation,
+  defs: &BTreeMap<String, Schema>,
+  extra_it: &mut BTreeMap<String, ObjectValidation>,
+) -> std::io::Result<()> {
+  if obj_valid.properties.len() > 0 {
+    writeln!(writer, "input {name} {{")?;
+    writer.indent();
+    for (name, property) in obj_valid.properties {
+      write_property(writer, name, property, defs, extra_it)?;
+    }
+    writer.unindent();
+    writeln!(writer, "}}")
+  } else {
+    Ok(())
+  }
+}
+
+fn write_all_input_types(
+  writer: &mut IndentedWriter<impl Write>,
+  mut extra_it: BTreeMap<String, ObjectValidation>,
+) -> std::io::Result<()> {
+  let schema = schemars::schema_for!(config::Config);
+
+  let defs = schema.definitions;
+  let mut scalars = vec![];
+  for (name, input_type) in defs.iter() {
+    let mut name = name.clone();
+    first_char_to_upper(&mut name);
+    write_input_type(
+      writer,
+      name,
+      input_type.clone().into_object(),
+      &defs,
+      &mut scalars,
+      &mut extra_it,
+    )?;
+  }
+
+  let mut new_extra_it = BTreeMap::new();
+
+  for (name, obj_valid) in extra_it.into_iter() {
+    write_object_validation(writer, name, obj_valid, &defs, &mut new_extra_it)?;
+  }
+
+  for name in scalars {
+    writeln!(writer, "scalar {name}")?;
+  }
+
+  Ok(())
+}
+
+fn generate_rc_file(file: File) -> Result<()> {
+  let mut file = IndentedWriter::new(file);
+  let mut written_directives = HashSet::new();
+  let wd = &mut written_directives;
+
+  let mut extra_it = BTreeMap::new();
+  let extra_it_ref = &mut extra_it;
+
+  write_schema_for::<config::Server>(&mut file, "Server", "SCHEMA", wd, extra_it_ref)?;
```

Comment:
can we use enums instead of string constants?

Reply:
converted them to enum